### PR TITLE
Add benchmark for json.compact high cpu usage in watch

### PR DESCRIPTION
**Pull request by mborsz**
#### What type of PR is this?

/kind feature

#### Which issue(s) this PR fixes:
Adds repro benchmark for https://github.com/kubernetes/kubernetes/issues/110146

It also does tiny refactor of existing benchmark, with no performance diff:
```
name               old time/op  new time/op  delta
WatchHTTP-48       31.5µs ± 0%  33.3µs ± 0%   ~     (p=1.000 n=1+1)
WatchHTTP_UTF8-48  31.9µs ± 0%  32.5µs ± 0%   ~     (p=1.000 n=1+1)
WatchWebsocket-48  32.9µs ± 0%  32.2µs ± 0%   ~     (p=1.000 n=1+1)
WatchProtobuf-48   18.6µs ± 0%  18.9µs ± 0%   ~     (p=1.000 n=1+1)
```

BenchmarkWatchCachingObjectJSON properly reproduces https://github.com/kubernetes/kubernetes/issues/110146:
image

Such branch isn't visible for BenchmarkWatchCachingObjectProtobuf:
image

#### Special notes for your reviewer:

It makes CachingObject public so that we can reuse it for performance testing. Alternative approach would be to reimplement some simpler mock, if prefered.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

/assign @wojtek-t (for comments on making CachingObject public)

Wojtek, is it OK to make this public or we prefer to implement fakeCachingObject in the watch_test.go?

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/endpoints/watch_test.go [hunk 1]
```diff
@@ -878,7 +879,7 @@ func BenchmarkWatchHTTP(b *testing.B) {
 		item.Name = fmt.Sprintf("reasonable-name-%d", i)
```
Comment (stevekuznetsov) [inline staging/src/k8s.io/apiserver/pkg/endpoints/watch_test.go:879]: nit: self-links were removed, may or may not be useful to remove these, perhaps in some follow-up

**Review by stevekuznetsov — COMMENTED** [review]
Benchmark refactors here look good.

/lgtm

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/caching_object.go [hunk 2]
```diff
@@ -56,12 +56,12 @@ type serializationResult struct {
...
 
-// cachingObject is an object that is able to cache its serializations
+// CachingObject is an object that is able to cache its serializations
 // so that each of those is computed exactly once.
 //
-// cachingObject implements the metav1.Object interface (accessors for
+// CachingObject implements the metav1.Object interface (accessors for
 // all metadata fields).
-type cachingObject struct {
+type CachingObject struct {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/caching_object.go:64]: I think I would prefer avoiding making this public, but I see that it's making this benchmark fairly simple to add.

@liggitt - WDYT about making this public?

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/caching_object.go:64]: I really don't want to make these public... this was a pretty messy implementation detail to begin with and I don't want any chance of accumulating external use

Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/caching_object.go:64]: +1 this can't be public

Comment (mborsz) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/caching_object.go]: I reimplemented simpler version.

**Comment by pacoxu** [comment]
/retest
for flake Test_Run_OneVolumeDetachFailNodeWithReadWriteOnce

**Pull request merged**